generator/src/main/java/org/apache/kafka/message/JsonConverterGenerator.java

```diff
@@ -380,8 +380,9 @@ private void generateVariableLengthTargetToJson(Target target, Versions versions
                         target.sourceVariable(), target.sourceVariable())));
             }
         } else if (target.field().type().isRecords()) {
-            headerGenerator.addImport(MessageGenerator.BINARY_NODE_CLASS);
-            buffer.printf("%s;%n", target.assignmentStatement("new BinaryNode(new byte[]{})"));
+            headerGenerator.addImport(MessageGenerator.INT_NODE_CLASS);
+            buffer.printf("%s;%n", target.assignmentStatement(
+                    String.format("new IntNode(%s.sizeInBytes())", target.sourceVariable())));
```

Comment:
I am not entirely convinced by this change. The aim of the auto-generated protocol is to serialize/deserialise the requests/responses. Thus, it is a little weird that we only retain the size when serializing records to JSON and we actually expect bytes when deserializing from JSON. We are not consistent.

I can think of the following alternatives:
1) We keep exceptions for both the produce request and the fetch response in `RequestConvertToJson` where we could add a `recordsSizeInBytes` field instead of the `records` field.
2) We could post-process the JSON tree where we could discard `records` and add `recordsSizeInBytes`.
3) We could extend the autogenerated protocol to do only retain the size based on a flag. I am thinking about a `verbose` flag like we had before.

I lean towards 3) at the moment. By default, we should keep the current behavior and only retains the size when `verbose` is `false`. We may find a better name than `verbose` though.

@anatasiavela @lbradstreet Would this make sense?

Reply:
@dajac if I understand correctly, the current generated JSON does not print the recordSet either. Is it breaking anything to return the size rather than an empty array?
```
            }
            if (_object.recordSet == null) {
                _node.set("recordSet", NullNode.instance);
            } else {
                _node.set("recordSet", new BinaryNode(new byte[]{}));
            }
```

That said, a verbose mode where we print out a byte array seems reasonable if we don't use it in the trace logging path by default?

Reply:
@lbradstreet The current generated JSON does not print the recordSet. When we were serializing a BinaryNode with an empty array, it was still being deserialized as a BinaryNode, so it doesn't break anything.
```
buffer.printf("%s;%n", target.assignmentStatement(
                String.format("MemoryRecords.readableRecords(ByteBuffer.wrap(MessageUtil.jsonNodeToBinary(%s, \"%s\")))",
                    target.sourceVariable(), target.humanReadableName())));
```
So I believe the concern is that I had proposed to change the serialization to an IntNode, but the deserialization still expects a BinaryNode which I should have also changed.
Another alternative I could fix this by changing the deserialization to expect an IntNode, but I'm for adding a verbose flag

Reply:
Nvm, we can't deserialize `recordSet` to an IntNode because `FetchResponseData` expects `recordSet` to be of type `BaseRecords`, not `int`.
What I've done was left deserialization the same and added the verbose tag to serialization.

Reply:
@lbradstreet No, it does not break anything but it is inconsistent. If we put our case aside, the purpose of the JSON converters is to convert to/from JSON so it is fair to expect that we have to maintain this contract by default. Changing it to return the size would break this contract.

Regarding `_node.set("recordSet", new BinaryNode(new byte[]{}));`, I do believe that this is a bug actually...

> That said, implementing a verbose mode (3) where we print out a byte array seems reasonable if we don't use it in the trace logging path by default?

Agreed. When logging requests/responses, we should stick to the non-verbose mode.